File: 03-PracticalImplementation/samples/javascript/mcp_sample.js
// MCP JavaScript/Node.js Sample Implementation

const EventEmitter = require('events');

/**
 * MCP Server implementation in JavaScript
 */
class MCPServer {
  constructor(options = {}) {
    this.serverName = options.serverName || 'JavaScript MCP Server';
    this.version = options.version || '1.0.0';
    this.models = options.models || ['gpt-4', 'llama-3-70b', 'claude-3-sonnet'];
    this.events = new EventEmitter();
  }

  /**
   * Handle an MCP completion request
   * @param {Object} request - The MCP request object
   * @returns {Promise<Object>} - The MCP response
   */
  async handleCompletionRequest(request) {
    console.log(`Processing completion request for model: ${request.model || 'unknown'}`);
    
    // Validate request
    if (!request.model) {
      return { error: 'Model is required' };
    }
    
    if (!this.models.includes(request.model)) {
      return { error: `Model ${request.model} not supported` };
    }
    
    if (!request.prompt) {
      return { error: 'Prompt is required' };
    }
    
    // Emit event for monitoring/metrics
    this.events.emit('request', { 
      type: 'completion', 
      model: request.model, 
      timestamp: new Date() 
    });
    
    // In a real implementation, this would call an AI model
    // Here we just echo back parts of the request with a mock response
    const response = {
      id: `mcp-resp-${request.id || Date.now()}`,
      model: request.model,
      response: `This is a response to: ${request.prompt.substring(0, 30)}...`,
      usage: {
        promptTokens: request.prompt.split(' ').length,
        completionTokens: 20,
        totalTokens: request.prompt.split(' ').length + 20
      }
    };
    
    // Simulate network delay
    await new Promise(resolve => setTimeout(resolve, 500));
    
    // Emit completion event
    this.events.emit('completion', {
      requestId: request.id,
      model: request.model,
      timestamp: new Date()
    });
    
    return response;
  }
  
  /**
   * Stream a completion response
   * @param {Object} request - The MCP request object
   * @returns {AsyncGenerator<Object>} - The streaming response
   */
  async *streamCompletion(request) {
    // Validate request
    if (!request.model || !request.prompt) {
      yield { error: 'Invalid request parameters' };
      return;
    }
    
    // Mock streaming response generation
    const words = `This is a streaming response to your prompt about ${
      request.prompt.split(' ').slice(0, 3).join(' ')
    }... The response will come word by word as a demonstration of streaming capabilities.`.split(' ');
    
    for (const word of words) {
      // Simulate token-by-token generation
      await new Promise(resolve => setTimeout(resolve, 100));
      yield {
        id: `mcp-stream-${request.id || Date.now()}`,
        model: request.model,
        delta: word + ' ',
        finished: false
      };
    }
    
    // Final message
    yield {
      id: `mcp-stream-${request.id || Date.now()}`,
      model: request.model,
      delta: '',
      finished: true,
      usage: {
        promptTokens: request.prompt.split(' ').length,
        completionTokens: words.length,
        totalTokens: request.prompt.split(' ').length + words.length
      }
    };
  }
}

/**
 * MCP Client implementation in JavaScript
 */
class MCPClient {
  constructor(serverUrl) {
    this.serverUrl = serverUrl;
  }
  
  /**
   * Send a completion request to an MCP server
   * @param {Object} options - The request options
   * @returns {Promise<Object>} - The completion response
   */
  async createCompletion(options) {
    console.log(`Sending completion request to ${this.serverUrl}`);
    
    if (!options.model || !options.prompt) {
      throw new Error('Model and prompt are required');
    }
    
    const request = {
      id: `req-${Date.now()}`,
      model: options.model,
      prompt: options.prompt,
      temperature: options.temperature ?? 0.7,
      maxTokens: options.maxTokens ?? 100
    };
    
    // In a real implementation, this would use fetch or axios to send an HTTP request
    // For this sample, we'll create a server and call it directly
    const server = new MCPServer();
    return await server.handleCompletionRequest(request);
  }
  
  /**
   * Stream a completion from an MCP server
   * @param {Object} options - The request options
   * @returns {AsyncGenerator<Object>} - The streaming response
   */
  async *createCompletionStream(options) {
    console.log(`Starting streaming completion with ${this.serverUrl}`);
    
    if (!options.model || !options.prompt) {
      throw new Error('Model and prompt are required');
    }
    
    const request = {
      id: `req-stream-${Date.now()}`,
      model: options.model,
      prompt: options.prompt,
      temperature: options.temperature ?? 0.7,
      maxTokens: options.maxTokens ?? 100,
      stream: true
    };
    
    // In a real implementation, this would use fetch with streaming or a WebSocket
    // For this sample, we'll create a server and yield from its generator
    const server = new MCPServer();
    yield* server.streamCompletion(request);
  }
}

// Example usage
async function runExample() {
  const client = new MCPClient('https://mcp.example.org');
  
  // Standard completion example
  console.log('\n--- Standard Completion Example ---');
  const response = await client.createCompletion({
    model: 'gpt-4',
    prompt: 'Explain the Model Context Protocol in simple terms',
    temperature: 0.7,
    maxTokens: 100
  });
  
  console.log(JSON.stringify(response, null, 2));
  
  // Streaming example
  console.log('\n--- Streaming Example ---');
  const streamingClient = new MCPClient('https://mcp.example.org');
  
  let fullResponse = '';
  for await (const chunk of streamingClient.createCompletionStream({
    model: 'gpt-4',
    prompt: 'Explain streaming in MCP',
    temperature: 0.7
  })) {
    if (chunk.delta) {
      process.stdout.write(chunk.delta);
      fullResponse += chunk.delta;
    }
    if (chunk.finished) {
      console.log('\n\nStream finished');
      console.log('Usage:', chunk.usage);
    }
  }
}

// Run the example if this file is executed directly
if (require.main === module) {
  runExample().catch(console.error);
}

module.exports = { MCPServer, MCPClient };

File: 03-PracticalImplementation/samples/python/mcp_sample.py
#!/usr/bin/env python
# MCP Python Sample Implementation

import asyncio
import json
import logging
from typing import Dict, List, Optional, Union

# Mock MCP server implementation
class MCPServer:
    def __init__(self, server_name: str, version: str):
        self.server_name = server_name
        self.version = version
        self.models = ["gpt-4", "llama-3-70b", "claude-3-sonnet"]
        self.logger = logging.getLogger("mcp.server")
        self.logger.setLevel(logging.INFO)
        handler = logging.StreamHandler()
        formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
        handler.setFormatter(formatter)
        self.logger.addHandler(handler)
    
    async def handle_completion_request(self, request: Dict) -> Dict:
        """Handle an MCP completion request"""
        self.logger.info(f"Processing completion request for model: {request.get('model', 'unknown')}")
        
        # Validate request
        if 'model' not in request:
            return {"error": "Model is required"}
        
        if request['model'] not in self.models:
            return {"error": f"Model {request['model']} not supported"}
            
        if 'prompt' not in request:
            return {"error": "Prompt is required"}
            
        # In a real implementation, this would call an AI model
        # Here we just echo back parts of the request with a mock response
        response = {
            "id": "mcp-resp-" + request.get('id', '12345'),
            "model": request['model'],
            "response": f"This is a response to: {request['prompt'][:30]}...",
            "usage": {
                "prompt_tokens": len(request['prompt'].split()),
                "completion_tokens": 20,
                "total_tokens": len(request['prompt'].split()) + 20
            }
        }
        
        # Simulate network delay
        await asyncio.sleep(0.5)
        return response

# MCP Client implementation
class MCPClient:
    def __init__(self, server_url: str):
        self.server_url = server_url
        self.logger = logging.getLogger("mcp.client")
        self.logger.setLevel(logging.INFO)
        
    async def create_completion(self, 
                               model: str, 
                               prompt: str, 
                               temperature: float = 0.7, 
                               max_tokens: int = 100) -> Dict:
        """Send a completion request to an MCP server"""
        self.logger.info(f"Sending completion request to {self.server_url}")
        
        # In a real implementation, this would use HTTP to send the request
        # Here we simulate by directly creating and returning a request object
        request = {
            "id": "req-123456",
            "model": model,
            "prompt": prompt,
            "temperature": temperature,
            "max_tokens": max_tokens
        }
        
        # This would normally be sent over HTTP
        # For this sample, we'll create a server and call it directly
        server = MCPServer("Sample MCP Server", "1.0")
        return await server.handle_completion_request(request)

async def main():
    # Example usage
    client = MCPClient("https://mcp.example.org")
    response = await client.create_completion(
        model="gpt-4",
        prompt="Explain the Model Context Protocol in simple terms",
        temperature=0.7,
        max_tokens=100
    )
    
    print("\n--- MCP Response ---")
    print(json.dumps(response, indent=2))

if __name__ == "__main__":
    asyncio.run(main())
